File: 5-Problem-A-Day/328. Odd Even Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode oddHead = new ListNode(0), evenHead = new ListNode(0);        
        ListNode oddPrev = oddHead, evenPrev = evenHead;
        ListNode curr = head;
        boolean even = false;
        while (curr != null) {
            ListNode tmp = curr.next;
            if (even) {
                evenPrev.next = curr;
                curr.next = null;
                evenPrev = curr;
            }
            else {
                oddPrev.next = curr;
                curr.next = null;
                oddPrev = curr;
            }
            even = !even;
            curr = tmp;
        }
        oddPrev.next = evenHead.next;
        return oddHead.next;
    }
}c

File: 2021/328. Odd Even Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }
        bool isOdd = true;
        ListNode *oddHead = head, *evenHead = head->next;
        ListNode *oddPrev = oddHead, *evenPrev = evenHead;
        head = evenHead->next;
        oddHead->next = NULL, evenHead->next = NULL;
        while (head != NULL) {
            if (isOdd) {
                oddPrev->next = head;
                oddPrev = head;
            } else {
                evenPrev->next = head;
                evenPrev = head;
            }
            head = head->next;
            isOdd = !isOdd;            
        }
        evenPrev->next = NULL;
        oddPrev->next = evenHead;
        return oddHead;
    }
};